File: agent-api/src/main/scala/kamon/api/instrumentation/KamonInstrumentation.scala
/*
 * =========================================================================================
 * Copyright © 2013-2016 the kamon project <http://kamon.io/>
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 * =========================================================================================
 */

package kamon.api.instrumentation

import java.lang.instrument.Instrumentation

import kamon.api.instrumentation.listener.InstrumentationListener
import kamon.api.instrumentation.mixin.{ MixinClassVisitorWrapper, MixinDescription }
import net.bytebuddy.agent.builder.AgentBuilder
import net.bytebuddy.agent.builder.AgentBuilder.Transformer
import net.bytebuddy.description.`type`.TypeDescription
import net.bytebuddy.description.method.MethodDescription
import net.bytebuddy.dynamic.DynamicType.Builder
import net.bytebuddy.matcher.ElementMatcher
import net.bytebuddy.matcher.ElementMatcher.Junction
import net.bytebuddy.matcher.ElementMatchers._
import net.bytebuddy.pool.TypePool

import scala.collection.mutable

abstract class KamonInstrumentation {
  private var elementMatcher: ElementMatcher[_ >: TypeDescription] = _
  private val mixins = mutable.ListBuffer[MixinDescription]()
  private val transformers = mutable.ListBuffer[Transformer]()

  val typePool = TypePool.Default.ofClassPath()
  val NotDeclaredByObject: Junction[MethodDescription] = not(isDeclaredBy(classOf[Object]))
  val NotTakesArguments: Junction[MethodDescription] = not(takesArguments(0))

  //TODO:configure listener
  def register(instrumentation: Instrumentation): Unit = {
    val builder = new AgentBuilder.Default()
      .withListener(InstrumentationListener())
      .`type`(elementMatcher)

    mixins.foreach { mixin ⇒
      builder
        .transform(withTransformer((builder, _) ⇒ builder.visit(MixinClassVisitorWrapper(mixin))))
        .installOn(instrumentation)
    }
    transformers.foreach(transformer ⇒ builder.transform(transformer).installOn(instrumentation))
  }

  def withTransformer(f: ⇒ (Builder[_], TypeDescription) ⇒ Builder[_]) = new Transformer {
    override def transform(builder: Builder[_], typeDescription: TypeDescription): Builder[_] = {
      f.apply(builder, typeDescription)
    }
  }

  def addTransformation(f: ⇒ (Builder[_], TypeDescription) ⇒ Builder[_]): Unit = transformers += withTransformer(f)
  def forTypes(f: ⇒ ElementMatcher[_ >: TypeDescription]): Unit = elementMatcher = f
  def forType(f: ⇒ ElementMatcher[_ >: TypeDescription]): Unit = forTypes(f)
  def forTargetType(f: ⇒ String): Unit = forType(named(f))
  def forSubtypeOf(f: ⇒ String): Unit = forType(isSubTypeOf(typePool.describe(f).resolve()).and(not(isInterface())))
  def addMixin(clazz: ⇒ Class[_]): Unit = mixins += MixinDescription(elementMatcher, clazz)
}

File: agent/src/main/java/kamon/api/instrumentation/KamonInstrumentation.java
package kamon.api.instrumentation;

import javaslang.Function2;
import javaslang.collection.List;
import javaslang.control.Option;
import kamon.api.instrumentation.listener.InstrumentationListener1;
import kamon.instrumentation.mixin.MixinDescription;
import kamon.instrumentation.mixin.MixinClassVisitorWrapper;
import net.bytebuddy.agent.builder.AgentBuilder;
import net.bytebuddy.agent.builder.AgentBuilder.Identified;
import net.bytebuddy.agent.builder.AgentBuilder.Transformer;
import net.bytebuddy.description.type.TypeDescription;
import net.bytebuddy.dynamic.DynamicType.Builder;
import net.bytebuddy.matcher.ElementMatcher;
import net.bytebuddy.pool.TypePool;

import java.lang.instrument.Instrumentation;
import java.util.function.Supplier;

import static net.bytebuddy.matcher.ElementMatchers.*;

public abstract class KamonInstrumentation {
    private Option<ElementMatcher> elementMatcher = Option.none();
    private List<MixinDescription> mixins = List.empty();
    private List<Transformer> transformers = List.empty();

    protected static final TypePool typePool = TypePool.Default.ofClassPath();
    protected static final ElementMatcher.Junction NotDeclaredByObject = not(isDeclaredBy(Object.class));
    protected static final ElementMatcher.Junction NotTakesArguments = not(takesArguments(0));

    public void register(Instrumentation instrumentation) {
        Identified agentBuilder = new AgentBuilder.Default()
                .withListener(new InstrumentationListener1())
                .type(elementMatcher.getOrElseThrow(() -> new RuntimeException("")));

        mixins.forEach(mixin ->
                agentBuilder
                        .transform((builder, typeDescription) -> builder.visit(new MixinClassVisitorWrapper(mixin)))
                        .installOn(instrumentation));

        transformers.forEach(transformer -> agentBuilder.transform(transformer).installOn(instrumentation));
    }

    private Transformer withTransformer(Function2<Builder, TypeDescription, Builder> f) {return f::apply;}

    public void addTransformation(Function2<Builder, TypeDescription, Builder> f) {
     transformers.append(withTransformer(f));
    }

    public void forTypes(Supplier<ElementMatcher> f) { elementMatcher = Option.of(f.get());}
    public void forType(Supplier<ElementMatcher> f) {forTypes(f);}
    public void forTargetType(Supplier<String> f) {forType((() -> named(f.get())));}
    public void forSubtypeOf(Supplier<String> f){forType(() -> isSubTypeOf(typePool.describe(f.get()).resolve()).and(not(isInterface())));}
    public void addMixin(Supplier<Class<?>> f) {mixins.append(MixinDescription.of(elementMatcher.get(),f.get()));}

 }
